### PR TITLE
Vulkan: Add ImgRefs to track usage of image subresorces

### PR DESCRIPTION
**Pull request by bjoeris**
## Description

This adds a new data structure `ImgRefs` to track usage of image subresources (aspect, mip level, array layer). `ImgRefs` initially has a single `FrameRefType` for the entire image, and splits this out into separate `FrameRefType`s for ecah aspect, mip levels and/or array layers as necessary.

This also adds `ImageRange`, which is used as a common structure to specify the portion image which was accessed.

### CODE REVIEW

File: renderdoc/driver/vulkan/vk_resources.h [hunk 1]
```diff
@@ -1006,6 +1007,224 @@ struct AttachmentInfo
...
+    for(int level = 0; level < splitLevelCount; ++level)
+    {
+      int otherLevel = other.areLevelsSplit ? level : 0;
+      for(int layer = 0; layer < splitLayerCount; ++layer)
+      {
+        int otherLayer = other.areLayersSplit ? layer : 0;
+        int index = (aspectIndex * splitLevelCount + level) * splitLayerCount + layer;
+        int otherIndex =
+            (otherAspectIndex * otherSplitLevelCount + otherLevel) * otherSplitLayerCount +
+            otherLayer;
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_resources.h]: Could this be using `SubresourceIndex`? The function is useful (I was going to suggest adding it until I noticed you already had it) but seems unused currently. It will be slightly less efficient by checking e.g. `areLevelsSplit` each call, but I would be surprised if that's a blip on a profile.

File: renderdoc/driver/vulkan/vk_resources.cpp [hunk 2]
```diff
@@ -2919,6 +2919,94 @@ VkImageAspectFlags FormatImageAspects(VkFormat fmt)
...
+}
+
+void ImgRefs::Split(bool splitAspects, bool splitLevels, bool splitLayers)
+{
+  int newSplitAspectCount = 1;
+  int oldSplitAspectCount = 1;
+  if(splitAspects || areAspectsSplit)
+  {
+    newSplitAspectCount = GetAspectCount();
+    if(splitAspects)
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_resources.cpp]: Should this be `if(areAspectsSplit)` instead?

**Review by baldurk — COMMENTED** [review]
I think this looks good - I've made a couple of comments where things could use clarification.

It would also be good to get tests in for this to cover the different edge cases and splitting triggers, but I'm far from perfect about that so you don't have to for this PR.

File: renderdoc/driver/vulkan/vk_resources.h (hunk 1 above)
Comment (bjoeris) [reply renderdoc/driver/vulkan/vk_resources.h]: Done

File: renderdoc/driver/vulkan/vk_resources.cpp (hunk 2 above)
Comment (bjoeris) [reply renderdoc/driver/vulkan/vk_resources.cpp]: Yes, good catch! Although `oldSplitAspectCount` isn't actually used, so I just deleted this.

**Review by bjoeris — COMMENTED** [review]
That typo you spotted made me nervous, so I went ahead and added some (non-exhaustive) tests.

**Review by baldurk — APPROVED** [review]
Could you add `imgrefs_tests.cpp` to `CMakeLists.txt`? Otherwise :+1:.
